### ISSUE TITLE
Killing containerd breaks interaction with running containers

### ISSUE THREAD

**Issue by cpuguy83** [moby/moby#36367]
**Description**

When containerd is killed/restarted, I am no longer able to interact with any running containers.

**Steps to reproduce the issue:**
1. `docker run -d --name test busybox top`
2. `pkill -9 docker-containerd`
3. wait for docker to restart containerd
4. `docker kill test`

**Describe the results you received:**

Receive this error in the logs: 
```
ERRO[2018-02-21T15:15:48.365281200Z] Handler for POST /v1.30/containers/test/kill returned error: Cannot kill container: test: Cannot kill container d542dc766a7b7e1a20d4c01a5b581339dfc84c46df691b2b0351fde082f0ecaa: grpc: the client connection is closing: failed precondition
```

If I manually kill the task with 
```
docker-containerd-ctr --namespace=moby --address /run/docker/containerd/docker-containerd.sock t kill <id> 
```

The event is observed in dockerd, so the client is active, but then dockerd tries to cleanup the container state in containerd and is unable to and gets the same error.

Tracking this down, it looks like an issue due to the way docker is caching container/task objects created from the old client. These objects have the client embedded in them and need to be recreated.

**Comment by cpuguy83** [comment]
Somewhat related to https://github.com/moby/moby/pull/36173 but the issue is that we are caching container and task objects from containerd which have the old client.

**Comment by boynux** [comment]
I can reproduce the issue on my machine however I don't think moby is the problem, using docker-containerd-ctr to kill the container, produces this output:

```
docker-containerd-ctr --namespace moby --address /run/docker/containerd/docker-containerd.sock t kill 82bfb7a25d414ced12dd9c6cc1ec0e64e2ae25adb6ace5d94e17bccfdafb80db
root@4f599fc31deb:/go/src/github.com/docker/docker# DEBU[0487] event published                               module="containerd/events" ns=moby topic="/tasks/exit" type=containerd.events.TaskExit
DEBU[2018-03-06T19:18:13.121047728Z] event                                         module=libcontainerd namespace=moby topic=/tasks/exit
DEBU[2018-03-06T19:18:13.122176064Z] DisableService 82bfb7a25d414ced12dd9c6cc1ec0e64e2ae25adb6ace5d94e17bccfdafb80db START
DEBU[2018-03-06T19:18:13.122210746Z] DisableService 82bfb7a25d414ced12dd9c6cc1ec0e64e2ae25adb6ace5d94e17bccfdafb80db DONE
DEBU[2018-03-06T19:18:13.122256523Z] Revoking external connectivity on endpoint test (b72a833f7665de43d9277ed7fbc5160b249529636c867ae9f3f8350f57f81265)
DEBU[2018-03-06T19:18:13.122797253Z] DeleteConntrackEntries purged ipv4:0, ipv6:0
DEBU[2018-03-06T19:18:13.316197005Z] Releasing addresses for endpoint test's interface on network bridge
DEBU[2018-03-06T19:18:13.316357524Z] ReleaseAddress(LocalDefault/172.18.0.0/16, 172.18.0.2)
DEBU[2018-03-06T19:18:13.485022649Z] Trying to unmount /var/lib/docker/containers/82bfb7a25d414ced12dd9c6cc1ec0e64e2ae25adb6ace5d94e17bccfdafb80db/mounts
DEBU[2018-03-06T19:18:13.525353574Z] Unmounted /var/lib/docker/containers/82bfb7a25d414ced12dd9c6cc1ec0e64e2ae25adb6ace5d94e17bccfdafb80db/mounts
ERRO[2018-03-06T19:18:13.525642829Z] 82bfb7a25d414ced12dd9c6cc1ec0e64e2ae25adb6ace5d94e17bccfdafb80db cleanup: failed to delete container from containerd: grpc: the client connection is closing: failed precondition
```

It eventually kills the container but the error is the same as `docker kill`

**Comment by cpuguy83** [comment]
The problem is we cache all the resources from containerd, which have a client embedded in the object without any way to swap it for a new client (and unfortunately grpc doesn't auto-reconnect).

containerd added a new client function `Reconnect` which would solve the issue here, unfortunately updating the containerd client can be tricky.

The other alternative is to re-populate the cache.

**Comment by cpuguy83** [comment]
This is fixed by b9b5dc37e37e67d1cd46d9a3448c96e3f57ef4bc
